File: console_edition_Java/MyChess/Zobrist.java
package MyChess;

import static MyChess.Config.BOARD_SIZE;

import java.util.Random;

import static MyChess.ChessType.*;

/**
 * Zobrist
 */
public class Zobrist {

    private long hum[][];
    private long com[][];
    private long zobristCode;
    private Random rnd;

    public Zobrist() {
        rnd = new Random();
        zobristCode = rnd.nextLong();
        hum = new long[BOARD_SIZE][BOARD_SIZE];
        com = new long[BOARD_SIZE][BOARD_SIZE];
        for (int x = 0; x < BOARD_SIZE; x++) {
            for (int y = 0; y < BOARD_SIZE; y++) {
                hum[x][y] = rnd.nextLong();
                com[x][y] = rnd.nextLong();
            }
        }
    }

    /**
     * 落子或撤销后更新哈希值
     * 
     * @param x      行号
     * @param y      列号
     * @param doType 棋子类型，只能是HUM_CHESS或COM_CHESS
     */
    public void goUpdate(int x, int y, int doType) {
        if (doType == HUM_CHESS) {
            zobristCode ^= hum[x][y];
        } else {
            zobristCode ^= com[x][y];
        }
    }

    /**
     * 获取棋盘的哈希值
     * 
     * @return 棋盘哈希值
     */
    public long code() {
        return zobristCode;
    }
}

File: src/GameAlgo/Zobrist.js
import Config from "./Config.js";
import { COM_CHESS, HUM_CHESS } from "./ChessType.js";

function randomNum() {
    return Math.floor(Math.random() * Number.MAX_SAFE_INTEGER);
}

/**
 * Zobrist
 */
export default class Zobrist {
    /**
     * @constructor
     */
    constructor() {
        /**
         * 空位状态
         * @type Number[]
         */
        this.empty = [];
        /**
         * 人类棋子状态
         * @type Number[]
         */
        this.hum = [];
        /**
         * 电脑棋子状态
         * @type Number[]
         */
        this.com = [];
        /**
         * 哈希值
         * @type Number
         */
        this.zobristCode = 0;

        for (let i = 0; i < Config.BOARD_SIZE * Config.BOARD_SIZE; i++) {
            this.empty.push(randomNum());
            this.hum.push(randomNum());
            this.com.push(randomNum());
        }
        for (let i = 0; i < Config.BOARD_SIZE * Config.BOARD_SIZE; i++) {
            this.zobristCode ^= this.empty[i];
        }
    }
    /**
     * 落子或撤销后更新哈希值
     * 
     * @param {Number} x 行号
     * @param {Number} y 列号
     * @param {Number} oldChess 旧棋子类型
     * @param {Number} newChess 新棋子类型
     */
    goUpdate(x, y, oldChess, newChess) {
        let index = x * Config.BOARD_SIZE + y;
        let oldArray, newArray;
        if (oldChess === HUM_CHESS) {
            oldArray = this.hum;
        } else if (oldChess === COM_CHESS) {
            oldArray = this.com;
        } else {
            oldArray = this.empty;
        }
        if (newChess === HUM_CHESS) {
            newArray = this.hum;
        } else if (newChess === COM_CHESS) {
            newArray = this.com;
        } else {
            newArray = this.empty;
        }
        this.zobristCode ^= oldArray[index];
        this.zobristCode ^= newArray[index];
    }
}